pd/src/app.rs

```diff
@@ -335,29 +333,106 @@ impl App {
                 .unwrap()
                 .add_transaction(verified_transaction);
 
-            increment_counter!("node_transactions_total");
+            metrics::increment_counter!("node_transactions_total");
             Ok(())
         }
     }
 
-    fn end_block(&mut self, end: EndBlock) -> response::EndBlock {
-        let epoch = self
+    fn end_block(&mut self, end: EndBlock) -> impl Future<Output = Result<Response, BoxError>> {
```

Comment:
The application logic for processing rate changes moved to the `EndBlock` handler; previously, it was included in the `State` code, but it's better for the state code to just handle writing state updates, rather than figuring out what the state updates should be.

Because processing rate changes requires querying the existing state, this method becomes async.